pkg/receiver/rawk8seventsreceiver/receiver.go

```diff
@@ -0,0 +1,239 @@
+// Copyright 2022, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package rawk8seventsreceiver
+
+import (
+	"context"
+	"errors"
+	"strings"
+	"time"
+
+	backoff "github.com/cenkalti/backoff/v4"
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/consumer/consumererror"
+	"go.opentelemetry.io/collector/model/pdata"
+	"go.uber.org/zap"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/runtime"
+	k8s "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+)
+
+// Only two types of events are created as of now.
+// For more info: https://docs.openshift.com/container-platform/4.9/rest_api/metadata_apis/event-core-v1.html
+var severityMap = map[string]pdata.SeverityNumber{
+	"normal":  pdata.SeverityNumberINFO,
+	"warning": pdata.SeverityNumberWARN,
+}
+
+type rawK8sEventsReceiver struct {
+	cfg              *Config
+	client           k8s.Interface
+	eventControllers []cache.Controller
+	eventCh          chan *corev1.Event
+	ctx              context.Context
+	cancel           context.CancelFunc
+	startTime        time.Time
+
+	consumer consumer.Logs
+	logger   *zap.Logger
+}
+
+// Interface for creating ListerWatcher objects. Used for injecting mocks into k8s informers.
+type ListerWatcherFactory func(c cache.Getter, resource string, namespace string, fieldSelector fields.Selector) cache.ListerWatcher
+
+func newRawK8sEventsReceiver(
+	params component.ReceiverCreateSettings,
+	cfg *Config,
+	consumer consumer.Logs,
+	client k8s.Interface,
+	listerWatcherFactory ListerWatcherFactory,
+) (*rawK8sEventsReceiver, error) {
+	var namespaceController cache.Controller
+	var namespaces []string
+
+	if listerWatcherFactory == nil {
+		listerWatcherFactory = func(
+			c cache.Getter,
+			resource string,
+			namespace string,
+			fieldSelector fields.Selector,
+		) cache.ListerWatcher {
+			return cache.NewListWatchFromClient(c, resource, namespace, fieldSelector)
+		}
+	}
```

Comment:
Is this because of test?

Reply:
I guess because of the call in `pkg/receiver/rawk8seventsreceiver/factory.go:78` 🤔 but I'd try to avoid this kind of thing.

Perhaps we could accept an `interface` instead and just use its method(s) then the code will get much simpler? And then you can use test factory implementation(s) in tests.

Reply:
`return newRawK8sEventsReceiver(params, rCfg, consumer, k8sClient, nil)`

we always pass `nil`, so this is not a "normal" behaviour to have `listerWatcherFactory`

Reply:
This is because of tests. The normal fake k8s client doesn't support the REST API, so in order to test informers, we need to mock the ListerWatcher - that's what the factory is for, and it's a factory because we need to potentially create one for each namespace we're watching. See `receiver_test.go` for how it's used.

The reason we need a wrapper here is that the return value of `ListerWatcherFactory` is `cache.ListerWatcher`, which is an interface, while `cache.NewListWatchFromClient` returns a `*cache.ListWatch`, which is a concrete type. Even though `cache.ListWatch` fulfills `cache.ListerWatcher`, these are different function types in Go.

@pmalek-sumo I'm not sure how we could make this simpler with an interface, can you elaborate.

Reply:
Please see: https://github.com/SumoLogic/sumologic-otel-collector/commit/9b80f40b03a5011efd58f811b0950c3d6886df91#diff-d9e51d1374cdf68c9bda647ff549bc33993e5ad1e16b774601d2c9cfc191955e

That's why I suggested the interface 🙃 If it's a func it's a different type but if it's an interface then it's ok to return a concrete type fulfilling the interface

Reply:
That's what I thought you meant, but in my view it's not really any simpler or more intuitive. My solution wraps the function type in another function, while yours wraps it in an interface. So you get simpler code at the cost of more complex types and more boilerplate.

With that said, I don't have a strong opinion on this. If your way is considered more idiomatic in Go, I can change to it.

Reply:
Depends on the perspective I guess. With `interface` approach you get the flexibility "at the expense" of an interface declaration. You can then pass anything you want to this function as long as it fulfils the interface and you separate the function body (as in the discussed snippet) from the actual "factory" implementation (which the caller **shouldn't** be bothered about IMHO).

Reply:
I added a comment explaining why the wrapper exists. I generally like just using function types as factories, so I'd prefer to keep this as is.

Reply:
The problem is that with this approach you have test specific code in production code.

With interface you have a customization point, without a need to mention that it's for mocks.

🤷

Reply:
Hm, that's a good point. I can avoid that by moving the wrapper into the factory, though. Sounds like a good idea in general.